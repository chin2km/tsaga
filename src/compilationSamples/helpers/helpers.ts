import * as path from 'path';
// tslint:disable-next-line:no-implicit-dependencies (Just used in tests)
import * as ts from 'typescript';

interface ReducedDiagnostic {
  message: string;
  fileName?: string;
  line?: number;
  character?: number;
}

export const defaultCompilerOptions: ts.CompilerOptions = {
  noEmit: true,
  strict: true,
  target: ts.ScriptTarget.ES2016,
  module: ts.ModuleKind.CommonJS,
  jsxFactory: 'react',
};

export function reducedDiagnostic(
  diagnostic: ts.Diagnostic,
): ReducedDiagnostic {
  if (diagnostic.file) {
    const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(
      diagnostic.start!,
    );
    const message = ts.flattenDiagnosticMessageText(
      diagnostic.messageText,
      '\n',
    );

    return {
      message,
      fileName: path.basename(diagnostic.file.fileName),
      line,
      character,
    };
  } else {
    return {
      message: ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n'),
    };
  }
}

export function getDiagnostics(
  fileNames: string[],
  options: ts.CompilerOptions,
): ReducedDiagnostic[] {
  const program = ts.createProgram(fileNames, options);
  const emitResult = program.emit();

  const diagnostics = ts
    .getPreEmitDiagnostics(program)
    .concat(emitResult.diagnostics);

  return diagnostics.map(reducedDiagnostic);
}

export function diagnosticsForFile(fileName: string) {
  return getDiagnostics([fileName], defaultCompilerOptions);
}
