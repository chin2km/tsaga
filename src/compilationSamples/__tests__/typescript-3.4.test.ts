import { diagnosticsForFile } from '../helpers/helpers';

test('Generic types are not correctly inferred (TS 3.4)', async () => {
  const diagnostics = diagnosticsForFile(`${__dirname}/../samples/typescript-3.4-inference.ts`);

  expect(diagnostics).toMatchSnapshot();

  // explicit alignment with comments in file, to catch accidental snapshot overwrite
  expect(diagnostics.length).toBe(2);
});
