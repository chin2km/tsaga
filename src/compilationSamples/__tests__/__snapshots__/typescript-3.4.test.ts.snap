// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic types are not correctly inferred (TS 3.4) 1`] = `
Array [
  Object {
    "character": 11,
    "fileName": "typescript-3.4-inference.ts",
    "line": 15,
    "message": "Property 'toString' does not exist on type 'never'.",
  },
  Object {
    "character": 2,
    "fileName": "typescript-3.4-inference.ts",
    "line": 21,
    "message": "Type '\\"string\\"' is not assignable to type 'number'.",
  },
]
`;
